File: shared/common/src/main/java/com/visor/school/common/api/GlobalExceptionHandler.java
package com.visor.school.common.api;

import jakarta.validation.ConstraintViolationException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.orm.ObjectOptimisticLockingFailureException;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;

import java.util.Map;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

@RestControllerAdvice
public class GlobalExceptionHandler {

    private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);

    private static String requestPath(WebRequest request) {
        String description = request.getDescription(false);
        return description.startsWith("uri=") ? description.substring(4) : description;
    }

    private ResponseEntity<ErrorResponse> buildResponse(
            ErrorCode errorCode,
            String message,
            HttpStatus status,
            WebRequest request
    ) {
        return ResponseEntity.status(status).body(
            ErrorResponse.of(errorCode, message, status, requestPath(request))
        );
    }

    private ResponseEntity<ErrorResponse> buildResponse(
            ErrorCode errorCode,
            String message,
            HttpStatus status,
            WebRequest request,
            Map<String, Object> details
    ) {
        return ResponseEntity.status(status).body(
            ErrorResponse.of(errorCode, message, status, requestPath(request), details)
        );
    }

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<ErrorResponse> handleIllegalArgumentException(
            IllegalArgumentException ex,
            WebRequest request
    ) {
        logger.warn("Illegal argument: {}", ex.getMessage(), ex);
        return buildResponse(
            ErrorCode.BAD_REQUEST,
            ex.getMessage() != null ? ex.getMessage() : "Invalid request",
            HttpStatus.BAD_REQUEST,
            request
        );
    }

    @ExceptionHandler(ConstraintViolationException.class)
    public ResponseEntity<ErrorResponse> handleConstraintViolationException(
            ConstraintViolationException ex,
            WebRequest request
    ) {
        logger.warn("Constraint violation: {}", ex.getMessage(), ex);
        Map<String, Object> details = ex.getConstraintViolations().stream()
                .collect(Collectors.toMap(v -> v.getPropertyPath().toString(), v -> (Object) v.getMessage()));
        return buildResponse(
            ErrorCode.VALIDATION_ERROR,
            Constants.ERROR_VALIDATION_FAILED,
            HttpStatus.BAD_REQUEST,
            request,
            details
        );
    }

    @ExceptionHandler(NoSuchElementException.class)
    public ResponseEntity<ErrorResponse> handleNoSuchElementException(
            NoSuchElementException ex,
            WebRequest request
    ) {
        logger.warn("Resource not found: {}", ex.getMessage());
        return buildResponse(
            ErrorCode.NOT_FOUND,
            ex.getMessage() != null ? ex.getMessage() : Constants.ERROR_RESOURCE_NOT_FOUND,
            HttpStatus.NOT_FOUND,
            request
        );
    }

    @ExceptionHandler(IllegalStateException.class)
    public ResponseEntity<ErrorResponse> handleIllegalStateException(
            IllegalStateException ex,
            WebRequest request
    ) {
        logger.warn("Illegal state: {}", ex.getMessage());
        return buildResponse(
            ErrorCode.INVALID_STATE,
            ex.getMessage() != null ? ex.getMessage() : Constants.ERROR_INVALID_STATE,
            HttpStatus.BAD_REQUEST,
            request
        );
    }

    @ExceptionHandler(ObjectOptimisticLockingFailureException.class)
    public ResponseEntity<ErrorResponse> handleOptimisticLockingFailure(
            ObjectOptimisticLockingFailureException ex,
            WebRequest request
    ) {
        logger.warn("Optimistic locking failure: {}", ex.getMessage());
        return buildResponse(
            ErrorCode.CONCURRENT_MODIFICATION,
            Constants.ERROR_CONCURRENT_MODIFICATION,
            HttpStatus.CONFLICT,
            request
        );
    }

    @ExceptionHandler(AccessDeniedException.class)
    public ResponseEntity<ErrorResponse> handleAccessDeniedException(
            AccessDeniedException ex,
            WebRequest request
    ) {
        logger.warn("Access denied: {}", ex.getMessage());
        return buildResponse(
            ErrorCode.ACCESS_DENIED,
            Constants.ERROR_ACCESS_DENIED,
            HttpStatus.FORBIDDEN,
            request
        );
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleGenericException(
            Exception ex,
            WebRequest request
    ) {
        logger.error("Unexpected error: {}", ex.getMessage(), ex);
        return buildResponse(
            ErrorCode.INTERNAL_ERROR,
            "An unexpected error occurred",
            HttpStatus.INTERNAL_SERVER_ERROR,
            request
        );
    }
}

File: services/user-service/src/main/kotlin/com/visor/school/user/exception/GlobalExceptionHandler.kt
package com.visor.school.user.exception

import com.visor.school.common.dto.ErrorDetail
import com.visor.school.common.exception.BusinessException
import com.visor.school.common.exception.ErrorResponse
import com.visor.school.common.exception.ResourceNotFoundException
import com.visor.school.common.exception.ValidationException
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import jakarta.servlet.http.HttpServletRequest

@RestControllerAdvice
class GlobalExceptionHandler {

    private val logger = LoggerFactory.getLogger(javaClass)

    @ExceptionHandler(ResourceNotFoundException::class)
    fun handleResourceNotFound(
        ex: ResourceNotFoundException,
        request: HttpServletRequest
    ): ResponseEntity<ErrorResponse> {
        logger.warn("Resource not found: {}", ex.message)
        return ResponseEntity.status(HttpStatus.NOT_FOUND)
            .body(ErrorResponse(
                status = HttpStatus.NOT_FOUND.value(),
                error = "Resource Not Found",
                message = ex.message ?: "Resource not found",
                path = request.requestURI
            ))
    }

    @ExceptionHandler(ValidationException::class)
    fun handleValidationException(
        ex: ValidationException,
        request: HttpServletRequest
    ): ResponseEntity<ErrorResponse> {
        logger.warn("Validation error: {}", ex.message)
        val errors = ex.errors?.map { ErrorDetail(it.key, it.value) } ?: emptyList()
        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
            .body(ErrorResponse(
                status = HttpStatus.BAD_REQUEST.value(),
                error = "Validation Error",
                message = ex.message ?: "Validation failed",
                path = request.requestURI,
                errors = errors
            ))
    }

    @ExceptionHandler(BusinessException::class)
    fun handleBusinessException(
        ex: BusinessException,
        request: HttpServletRequest
    ): ResponseEntity<ErrorResponse> {
        logger.warn("Business error: {}", ex.message)
        val status = when (ex.code) {
            "CONFLICT" -> HttpStatus.CONFLICT
            "UNAUTHORIZED" -> HttpStatus.UNAUTHORIZED
            else -> HttpStatus.BAD_REQUEST
        }
        return ResponseEntity.status(status)
            .body(ErrorResponse(
                status = status.value(),
                error = status.reasonPhrase,
                message = ex.message ?: "Business error occurred",
                path = request.requestURI
            ))
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleMethodArgumentNotValid(
        ex: MethodArgumentNotValidException,
        request: HttpServletRequest
    ): ResponseEntity<ErrorResponse> {
        logger.warn("Validation error: {}", ex.message)
        val errors = ex.bindingResult.fieldErrors.map { fieldError: FieldError ->
            ErrorDetail(fieldError.field, fieldError.defaultMessage ?: "Invalid value")
        }
        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
            .body(ErrorResponse(
                status = HttpStatus.BAD_REQUEST.value(),
                error = "Validation Error",
                message = "Validation failed",
                path = request.requestURI,
                errors = errors
            ))
    }

    @ExceptionHandler(Exception::class)
    fun handleGenericException(
        ex: Exception,
        request: HttpServletRequest
    ): ResponseEntity<ErrorResponse> {
        logger.error("Unexpected error: {}", ex.message, ex)
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
            .body(ErrorResponse(
                status = HttpStatus.INTERNAL_SERVER_ERROR.value(),
                error = "Internal Server Error",
                message = "An unexpected error occurred",
                path = request.requestURI
            ))
    }
}

